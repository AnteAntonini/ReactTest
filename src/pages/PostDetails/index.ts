import { lazy } from "react";

export const PostDetails = lazy(() => import("./PostDetails"));
