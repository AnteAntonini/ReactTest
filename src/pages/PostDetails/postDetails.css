.back-btn-wrapper {
  max-width: 800px;
  width: 80%;
}

.back-btn {
  display: inline-flex;
  align-items: center;
  text-decoration: none;
  color: var(--color-martian-dark);
}

.arrow-left {
  margin-right: 0.5rem;
}

.back-btn:hover .arrow-left {
  transform: scaleX(1.25);
}
