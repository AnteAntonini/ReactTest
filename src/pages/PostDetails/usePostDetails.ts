import { useParams } from "react-router-dom";
import { useFetch } from "../../hooks/useFetch";
import { TPostsResponse } from "../../types/TPosts";
import { TPostsCommentsResponse } from "../../types/TComments";
import { useEffect } from "react";
import { TUsersResponse } from "../../types/TUsers";

export const usePostDetails = (message: string) => {
  const { id: postId } = useParams();

  useEffect(() => {
    console.log(`${message} Posts Details`);
  }, []);

  const users: TUsersResponse[] = useFetch(
    "https://demo.martian.services/api/users"
  );

  const userInfo = users.find((user) => user.id === Number(postId));
  const userName = userInfo ? userInfo.name : undefined;

  const post: TPostsResponse = useFetch(
    `https://demo.martian.services/api/posts/${postId}`
  );

  const comments: TPostsCommentsResponse[] = useFetch(
    `https://demo.martian.services/api/comments?postId=${postId}`
  );
  return {
    post,
    comments,
    userName,
  };
};
