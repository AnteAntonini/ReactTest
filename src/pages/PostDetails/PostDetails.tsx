import { Link } from "react-router-dom";
import { ArrowLeft } from "../../assets";
import { usePostDetails } from "./usePostDetails";
import PostItem from "../../components/PostItem/PostItem";
import "./postDetails.css";

const PostDetails = ({ message }: { message: string }) => {
  const { post, comments, userName } = usePostDetails(message);

  return (
    <div className="flex-col-center">
      <h1 className="page-title">Posts Details</h1>

      <div className="back-btn-wrapper">
        <Link to="/app" className="back-btn">
          <ArrowLeft className="arrow-left" /> Go Back
        </Link>
      </div>
      {post && (
        <PostItem
          message="Hello from"
          id={post.id}
          postTitle={post.title}
          postBody={post.body}
          postUserName={userName}
          comments={comments}
          showCommentBody={true}
        />
      )}
    </div>
  );
};

export default PostDetails;
