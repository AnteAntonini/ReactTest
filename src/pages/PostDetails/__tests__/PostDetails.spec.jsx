import { render, screen } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import PostDetails from "../PostDetails";
import { usePostDetails } from "../usePostDetails";

// Mock the usePostDetails hook
jest.mock("../usePostDetails");

describe("PostDetails Component", () => {
  it("renders post details and a 'Go Back' button", () => {
    // Mock the usePostDetails hook return values
    usePostDetails.mockReturnValue({
      post: {
        id: 1,
        title: "Sample Post",
        body: "This is the post body.",
      },
      comments: [
        {
          id: 1,
          email: "user@example.com",
          name: "User",
          body: "This is a comment",
        },
      ],
      userName: "User123",
    });

    render(
      <Router>
        <PostDetails message="Test" />
      </Router>
    );

    // Check if the post details and 'Go Back' button are present
    expect(screen.getByText("Posts Details")).toBeInTheDocument();
    expect(screen.getByText("Sample Post")).toBeInTheDocument();
    expect(screen.getByText("This is the post body.")).toBeInTheDocument();
    expect(screen.getByText("User123")).toBeInTheDocument();
    expect(screen.getByText("This is a comment")).toBeInTheDocument();
    expect(screen.getByText("Go Back")).toBeInTheDocument();
  });

  it("displays post details with no comments", () => {
    // Mock the usePostDetails hook return values with no comments
    usePostDetails.mockReturnValue({
      post: {
        id: 2,
        title: "Post with No Comments",
        body: "This post has no comments.",
      },
      comments: [],
      userName: "User456",
    });

    render(
      <Router>
        <PostDetails message="Test" />
      </Router>
    );

    // Check if the post details and 'Go Back' button are present
    expect(screen.getByText("Posts Details")).toBeInTheDocument();
    expect(screen.getByText("Post with No Comments")).toBeInTheDocument();
    expect(screen.getByText("This post has no comments.")).toBeInTheDocument();
    expect(screen.getByText("User456")).toBeInTheDocument();

    // Ensure that there are no comment elements
    expect(screen.queryByText("No Comments")).toBeNull();
  });
});

describe("Test snapshot", () => {
  it("Component should match snapshot", () => {
    usePostDetails.mockReturnValue({
      post: {
        id: 2,
        title: "Post with No Comments",
        body: "This post has no comments.",
      },
      comments: [],
      userName: "User456",
    });

    const { asFragment } = render(
      <Router>
        <PostDetails message="Test" />
      </Router>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
