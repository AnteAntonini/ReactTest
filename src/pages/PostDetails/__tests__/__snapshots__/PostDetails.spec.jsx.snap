// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test snapshot Component should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex-col-center"
  >
    <h1
      class="page-title"
    >
      Posts Details
    </h1>
    <div
      class="back-btn-wrapper"
    >
      <a
        class="back-btn"
        href="/app"
      >
        <svg
          class="arrow-left"
        >
          arrow-left.svg
        </svg>
         Go Back
      </a>
    </div>
    <div
      class="post-section undefined"
    >
      <div
        class="post-title"
      >
        Post with No Comments
      </div>
      <div
        class="post-body"
      >
        This post has no comments.
      </div>
      <span
        class="post-username"
      >
        User456
      </span>
      <hr />
      <div>
        <h3
          class="comment-title"
        >
           0 Comments
        </h3>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
